bin/main

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/PerceivedComplexity
+
+# rubocop:disable Metrics/CyclomaticComplexity
+
+# rubocop:disable Metrics/ModuleLength
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+    entry
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+    entry
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    i = 0
+    arr = []
+    while i < size
+      arr << self[i] if yield(self[i])
+      i += 1
+    end
+    arr
+  end
+
+  def my_all?(pattern = nil)
+    unless block_given?
+      return my_all? { |obj| obj } unless pattern
+
+      if pattern.class == Regexp
+        my_all? { |x| pattern.match(x) }
+      elsif pattern.class == Class
+        my_all? { |x| x.is_a? pattern }
+      else
+        my_all? { |x| x === pattern }
+      end
+    end
+    my_each { |x| return false unless yield x }
+    true
+  end
```

Comment:
* [ ] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all?(Integer) == array.all?(Integer) # false
  ```
* [ ] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
  ```ruby
    array = %w[dog door rod blade]
    p array.my_all?(/c/) == array.all?(/c/) # false
  ```
* [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_all?(5) == array.all?(5) # false
  ```

- The method should take into consideration the case where it's called without a block. `unless block_given?` ignores that case.

Reply:
I honestly don't know any other way to solve this.